aten/src/ATen/cuda/CUDAFuture.cpp

```diff
@@ -91,19 +91,48 @@ std::string formatSetOfDevices(
   return oss.str();
 }
 
+const c10::impl::DeviceGuardImplInterface* getImplForDevices(
+    const std::vector<c10::Device>& devices) {
+  if (devices.empty()) {
+    return nullptr;
+  }
+  c10::DeviceType deviceType = devices[0].type();
+  for (size_t idx = 1; idx < devices.size(); idx++) {
+    TORCH_CHECK_VALUE(
+        devices[idx].type() == deviceType,
+        "Expected all devices to be of the same type, but got a mismatch between ",
+        devices[0],
+        " and ",
+        devices[idx]);
+  }
+  return c10::impl::getDeviceGuardImpl(deviceType);
+}
+
 // We need devices to be sorted in order to use set_difference.
-std::vector<c10::DeviceIndex> sortDevices(
-    std::vector<c10::DeviceIndex> devices) {
-  std::sort(devices.begin(), devices.end());
-  return devices;
+std::vector<c10::DeviceIndex> getSortedIndicesOfDevices(
+    const c10::impl::DeviceGuardImplInterface* impl,
+    const std::vector<c10::Device>& devices) {
+  std::vector<bool> isDeviceUsed(impl->deviceCount(), false);
+  for (const c10::Device& device : devices) {
+    TORCH_CHECK_VALUE(
+        device.has_index(), "Expected devices to have indices, got ", device);
+    isDeviceUsed[device.index()] = true;
+  }
+  std::vector<c10::DeviceIndex> deviceIndices;
+  for (c10::DeviceIndex idx = 0; idx < isDeviceUsed.size(); idx++) {
+    if (isDeviceUsed[idx]) {
+      deviceIndices.push_back(idx);
+    }
+  }
+  return deviceIndices;
 }
 
 } // namespace
 
-CUDAFuture::CUDAFuture(at::TypePtr type, std::vector<c10::DeviceIndex> devices)
+CUDAFuture::CUDAFuture(at::TypePtr type, std::vector<c10::Device> devices)
     : at::ivalue::Future(std::move(type)),
-      impl_(c10::impl::getDeviceGuardImpl(c10::kCUDA)),
-      devices_(sortDevices(std::move(devices))) {
+      impl_(getImplForDevices(devices)),
+      devices_(getSortedIndicesOfDevices(impl_, devices)) {
```

Comment:
curious, any reason for not just storing devices instead of device indices in this change??

Reply:
At this stage I wanted to change as few things as possible to keep the delta small and thus help see that this is not changing behaviors. But, yes, I was thinking the same and I'll probably do a follow-up "polishing" PR.